mllib/src/main/scala/org/apache/spark/mllib/util/MLUtils.scala

```diff
@@ -248,6 +248,19 @@ object MLUtils extends Logging {
     }.toArray
   }
 
+  /**
+   * Version of `kFold()` taking a fold column name.
+   */
+  @Since("3.1.0")
+  def kFold(df: DataFrame, numFolds: Int, foldColName: String): Array[(RDD[Row], RDD[Row])] = {
+    val foldCol = df.col(foldColName)
+    val dfWithMod = df.withColumn(foldColName, pmod(foldCol, lit(numFolds)))
+    (0 until numFolds).map { fold =>
+      (dfWithMod.filter(col(foldColName) =!= fold).drop(foldColName).rdd,
+        dfWithMod.filter(col(foldColName) === fold).drop(foldColName).rdd)
```

Comment:
I have a question here: 
If foldCol doesn't contain a certain specific fold num, the validation dataset for that iteration will be empty. For example, if numFolds is 3 and foldCol only contains 0 and 2, the validation dataset is empty for fold 1. Should we check for empty splits and remove them?

Reply:
Generally I think a valid user-specified folds should not miss any fold number. I mentioned it in above discussion that adding fold number check is possible however I don't want to do it because performance concern. I'd more towards letting users to be responsible for it if they want to specify fold numbers by themselves.

Reply:
I also perfer adding a numFolds checking here, but not strongly.
Since ML imlps tends to transform input dataframe to `RDD[Vector]` and then cache it, compared with the training, this checking maybe cheap.

Reply:
@zhengruifeng Is the check you suggested the same as @huaxingao? I.e., checking for empty datasets. Or checking user-specified fold numbers in the range [0, numFolds)?

For the later, now I take the value mod numFolds, and I think it should be enough for valid fold numbers.

Reply:
Both, 1, checking `foldCol` valuse are in [0, numFolds); 2, for each fold, both the train rdd and the validation rdd are not empty.

I am afraid if the `numFolds` is wrongly set, for example, numFolds=3, and the `foldCol` values are in {0, 1, 2, 3}, then there maybe task skewness;

Reply:
Though I think users should be confident enough to use user-specified fold numbers. :) But it also sounds good to me. I will add some checks in next commit.